### ISSUE TITLE
Thread panicked at 'cannot access a TLS value during or after it is destroyed: AccessError'

### ISSUE THREAD

**Issue by mentegy** [alexcrichton/curl-rust#226]
App written in C has some bindings to my rust lib. When C app receives SIGINT it calls the function (using c `atexit`) which perform a http request, but the threads is panicking:
```
thread '<unnamed>' panicked at 'cannot access a TLS value during or after it is destroyed: AccessError', libcore/result.rs:945:5
stack backtrace:
   0: std::sys::unix::backtrace::tracing::imp::unwind_backtrace
             at libstd/sys/unix/backtrace/tracing/gcc_s.rs:49
   1: std::sys_common::backtrace::print
             at libstd/sys_common/backtrace.rs:71
             at libstd/sys_common/backtrace.rs:59
   2: std::panicking::default_hook::{{closure}}
             at libstd/panicking.rs:211
   3: std::panicking::default_hook
             at libstd/panicking.rs:227
   4: std::panicking::rust_panic_with_hook
             at libstd/panicking.rs:511
   5: std::panicking::continue_panic_fmt
             at libstd/panicking.rs:426
   6: rust_begin_unwind
             at libstd/panicking.rs:337
   7: core::panicking::panic_fmt
             at libcore/panicking.rs:92
   8: core::result::unwrap_failed
   9: <std::thread::local::LocalKey<T>>::with
  10: curl::easy::handler::header_cb
  11: Curl_client_chop_write
  12: Curl_http_readwrite_headers
  13: Curl_readwrite
  14: multi_runsingle
  15: curl_multi_perform
  16: curl_easy_perform
  17: curl::easy::handle::Transfer::perform
```
I suspect that SIGINT closes some rust curl resources and then the code within `atexit` cannot do http requests anymore. Is there anyway to avoid a thread panic and perform a http request on  `atexit` ?

**Comment by mentegy** [comment]
We've tried to call `curl::init()` before `atexit` (maybe it is hooks order issue) but didn't have luck

**Comment by alexcrichton** [comment]
Hm the problem here looks like it's unrelated to curl, but perhaps local usage of thread local storage? It looks like thread local storage is in use while the thread is being destroyed?

**Comment by mentegy** [comment]
@alexcrichton Hello, thanks for a response, I've made a repo example:

```rust
extern crate curl;
extern crate libc;

use curl::easy::Easy;

pub extern "C" fn hook() {
    let mut easy = Easy::new();
    easy.url("google.com").unwrap();
    easy.write_function(|data| {
        Ok(data.len())
    }).unwrap();
    easy.perform().unwrap();
}

fn main() {
    curl::init();
    unsafe { libc::atexit(hook); }
    println!("Finishing...")
}
```

**Comment by mentegy** [comment]
Output:
```
Finishing...
thread '<unnamed>' panicked at 'cannot access a TLS value during or after it is destroyed: AccessError', libcore/result.rs:945:5
stack backtrace:
   0: std::sys::unix::backtrace::tracing::imp::unwind_backtrace
             at libstd/sys/unix/backtrace/tracing/gcc_s.rs:49
   1: std::sys_common::backtrace::print
             at libstd/sys_common/backtrace.rs:71
             at libstd/sys_common/backtrace.rs:59
   2: std::panicking::default_hook::{{closure}}
             at libstd/panicking.rs:211
   3: std::panicking::default_hook
             at libstd/panicking.rs:227
   4: std::panicking::rust_panic_with_hook
             at libstd/panicking.rs:463
   5: std::panicking::begin_panic_fmt
             at libstd/panicking.rs:350
   6: rust_begin_unwind
             at libstd/panicking.rs:328
   7: core::panicking::panic_fmt
             at libcore/panicking.rs:71
   8: core::result::unwrap_failed
   9: <std::thread::local::LocalKey<T>>::with
  10: curl::easy::handler::opensocket_cb
  11: <unknown>
  12: <unknown>
  13: <unknown>
  14: <unknown>
  15: <unknown>
  16: <unknown>
  17: curl_multi_perform
  18: curl_easy_perform
  19: curl::easy::handle::Easy::perform
  20: curl_test::hook
```

I think that this can be resolved if resources need to perform request could be re-initiated in shutdown hook, but how to achieve that?

**Comment by alexcrichton** [comment]
Ah right yeah we've got some TLS! Should be fixed now, thanks for the reproduction!

**Comment by mentegy** [comment]
@alexcrichton Wow, thanks, that was quick! Is there a release version which contains the fix?

**Comment by alexcrichton** [comment]
I've published now! Should be in 0.4.16

**Comment by mentegy** [comment]
@alexcrichton Thanks a lot, worked for us!
